flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/ttl/RocksDbTtlCompactFilterUtils.java

```diff
@@ -0,0 +1,179 @@
+			}
+		}
+		return null;
+	}
+
+	private static org.rocksdb.Logger createRocksDbNativeLogger() {
+		if (LOG.isDebugEnabled()) {
+			// options are always needed for org.rocksdb.Logger construction (no other constructor)
+			// the logger level gets configured from the options in native code
+			try (DBOptions opts = new DBOptions().setInfoLogLevel(InfoLogLevel.DEBUG_LEVEL)) {
+				return new org.rocksdb.Logger(opts) {
+					@Override
+					protected void log(InfoLogLevel infoLogLevel, String logMsg) {
+						LOG.debug("RocksDB filter native code log: " + logMsg);
+					}
+				};
+			}
+		} else {
+			return null;
+		}
+	}
+
+	public void configCompactFilter(
+			@Nonnull StateDescriptor<?, ?> stateDesc, FlinkCompactionFilter compactionFilter) {
+		StateTtlConfig ttlConfig = stateDesc.getTtlConfig();
+		boolean useSystemTime = !useCustomTimeForTtlCompactFilter &&
+			ttlConfig.getTimeCharacteristic() == StateTtlConfig.TimeCharacteristic.ProcessingTime;
+		configCompactFilter(stateDesc, compactionFilter, useSystemTime);
+	}
+
+	private void configCompactFilter(
+			@Nonnull StateDescriptor<?, ?> stateDesc, FlinkCompactionFilter compactionFilter, boolean useSystemTime) {
+		StateTtlConfig ttlConfig = stateDesc.getTtlConfig();
+		if (ttlConfig.isEnabled() && ttlConfig.getCleanupStrategies().inRocksdbCompactFilter()) {
+			if (!enableTtlCompactionFilter) {
+				LOG.warn("Cannot configure RocksDB TTL compaction filter for state <{}>: " +
+					"feature is disabled for the state backend.", stateDesc.getName());
+			}
+			assert compactionFilter != null;
+			FlinkCompactionFilter.Config config;
+			if (stateDesc instanceof ListStateDescriptor) {
+				ListStateDescriptor<?> listStateDesc = (ListStateDescriptor<?>) stateDesc;
+				int len = listStateDesc.getElementSerializer().getLength();
+				if (len > 0) {
+					config = FlinkCompactionFilter.Config.createForFixedElementList(
+						0, ttlConfig.getTtl().toMilliseconds(), useSystemTime, len + 1);
+				} else {
+					config = FlinkCompactionFilter.Config.createForList(
+						0, ttlConfig.getTtl().toMilliseconds(), useSystemTime,
+						new ListElementIter<>(listStateDesc.getElementSerializer()));
+				}
+			} else {
+				int timestampOffset = stateDesc instanceof MapStateDescriptor ? 1 : 0;
+				config = FlinkCompactionFilter.Config.create(
+					FlinkCompactionFilter.StateType.Value, timestampOffset, ttlConfig.getTtl().toMilliseconds(), useSystemTime);
+			}
+			compactionFilter.configure(config);
+		}
+	}
+
+	private static class ListElementIter<T> implements FlinkCompactionFilter.ListElementIter {
+		private final TypeSerializer<T> serializer;
+		private DataInputDeserializer input;
+
+		private ListElementIter(TypeSerializer<T> serializer) {
+			this.serializer = serializer;
+		}
+
+		@Override
+		public void setListBytes(byte[] bytes) {
+			input = new DataInputDeserializer(bytes);
+		}
+
+		@Override
+		public int nextOffset(int currentOffset) {
+			try {
+				serializer.deserialize(input);
```

Comment:
A little question: If there are serval compaction threads, will they share the`ListElementIter` object here to lead to the race condition?

Reply:
Good point, earlier they were stateless but after optimising and setting the byte value of list before iterating the elements, it can be accessed concurrently. I refactored the code to use factories which produce objects used by single thread.

Reply:
Thanks for your confirm. I will take a look at this change. And @StefanRRichter do you have any suggestion on this feature/implementation.  (*I want to use this feature in our inner flink for production, so i want to confirm the validity* )